### PR TITLE
feat: interval add timestamp

### LINKED ISSUE #5413: Interval plus(+) Timestamp operation support
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
While we can add intervals to timestamp variants (LHS: Timestamp (like `TimestampSecond`), RHS: Interval (like `IntervalYearMonth`)), we cannot replace the sides (LHS: Interval, RHS: Timestamp), and get the error 
`Err(Internal("Operator + is not implemented for variants IntervalYearMonth(\"1\") and TimestampSecond(12345678, None)"))`
Addition should be commutative for intervals and timestamps, unlike subtraction.

**Describe the solution you'd like**
In `impl_op` macro, the match arms can be extended to handle Interval variants for addition, and an adder function can be implemented.

**Describe alternatives you've considered**
Only the order of the parameters can be changed before calling that `impl_op` macro since **Timestamp + Interval** operations are working.

**Additional context**
This feature can also be extended to cover operations between interval variants (`IntervalYearMonth`, `IntervalDayTime`, `IntervalMonthDayNano`). We can use addition and subtraction operators between them (negatively resulting situations should be under consideration).

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #5413 

# Rationale for this change
As the `timestamp` + feature has already been implemented, you can further enhance the `imp_op` function by adding additional matching conditions to support `interval` operations.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 1]
```diff
@@ -191,3 +191,24 @@ query P
 SELECT DATE_TRUNC('second', '2022-08-03 14:38:50Z');
 ----
 2022-08-03T14:38:50
+
+# Test that interval can add a timestamp
```
Comment (korowa) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:195]: Thank you!

I also suppose it's worth adding tests for `interval + timestamp` case, as it was added in scalar.rs

File: datafusion/physical-expr/src/planner.rs [hunk 2]
```diff
@@ -197,6 +197,16 @@ pub fn create_physical_expr(
                     rhs,
                     input_schema,
                 )?)),
+                (
+                    DataType::Interval(_),
```
Comment (alamb) [inline datafusion/physical-expr/src/planner.rs:201]: 👍

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:195]: I see tests here for `timestamp + interval` and `interval + timestamp` (e.g. `SELECT INTERVAL '8' DAY + timestamp '2013-07-01 12:00:00';`)

I don't see tests for `date + timestamp` and `timestamp + date` -- maybe that is what you meant?

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H  -- this is awesome. I'll wait for your response to @korowa 's comment.

**Comment by Weijun-H** [comment]
Hi @alamb, I have not tested `date + timestamp` and `timestamp + date` because they are invalid computations. Do you mean that I should test `date + interval` and `interval + date` instead?

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt (hunk 1 above)
Comment (korowa) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:195]: I meant last four tests, added in 2nd commit

**Review by korowa — APPROVED** [review]

**Comment by korowa** [comment]
LGTM

Thanks again @Weijun-H!

**Comment by alamb** [comment]
>  Do you mean that I should test date + interval and interval + date instead?

No sorry for not being clear. I was confused about @korowa 's suggestion. No changes needed. Thank you!

**Pull request merged**